ftdc/cmd/parser.go

```diff
@@ -132,10 +132,197 @@ func (gpw *gnuplotWriter) addPoint(timeSeconds int64, metricName string, metricV
+	// the necessary numerators and denominators. The second pass will write the computed datapoint
+	// to the underlying gnuplotWriter.
+	//
+	// Ratio metrics are identified by the metric suffix. E.g: `rdk.custom_module.UserCPUSecs` will
+	// be classified as a (numerator in a) ratio metric. We must also take care to record the prefix
+	// of the ratio metric, the "metric identifier". There may be `rdk.foo_module.UserCPUSecs` in
+	// addition to `rdk.bar_modular.UserCPUSecs`. Which should create two CPU% graphs.
 	for _, reading := range datum.Readings {
+		// pullRatios will identify if the metric is a "ratio" metric. If so, we do not currently
+		// know what to graph and `pullRatios` will accumulate the relevant information into
+		// `deferredReadings`.
+		isRatioMetric := pullRatios(reading, datum.ConvertedTime().Unix(), ratioMetricToFields, deferredReadings)
+		if isRatioMetric {
+			// Ratio metrics need to be compared to some prior ratio metric to create a data
+			// point. We do not output any information now. We instead accumulate all of these
+			// results to be later used. These are named "deferred values".
+			continue
+		}
+
 		gpw.addPoint(datum.ConvertedTime().Unix(), reading.MetricName, reading.Value)
 	}
+
+	return deferredReadings
+}
+
+// Ratios are averaged over a "recent history". This window size refers to a time in seconds, but we
+// actually measure with respect to consecutive FTDC readings. The output value will use the system
+// clock difference to compute a correct rate. We just accept there may be fuzziness with respect to
+// how recent of a history we're actually using.
+//
+// Consider adding logging when two FTDC readings `windowSizeSecs` apart is not reflecting by their
+// system time difference.
+const windowSizeSecs = 5
+
+// The deferredValues input is in FTDC reading order. On a responsive system, adjacent items in the
+// slice should be one second apart.
+func (gpw *gnuplotWriter) writeDeferredValues(deferredValues []map[string]*ratioReading, logger logging.Logger) {
+	for idx, currReadings := range deferredValues {
+		if idx == 0 {
+			// The first element cannot be compared to anything. It would create a divide by zero
+			// problem.
+			continue
+		}
+
+		// `forCompare` is the index element to compare the "current" element pointed to by `idx`.
+		forCompare := idx - windowSizeSecs
+		if forCompare < 0 {
+			// If we haven't
+			forCompare = 0
+		}
+
+		prevReadings := deferredValues[forCompare]
+		for metricName, currRatioReading := range currReadings {
+			var diff ratioReading
+			if prevratioReading, exists := prevReadings[metricName]; exists {
+				diff = currRatioReading.diff(prevratioReading)
+			} else {
+				logger.Infow("Deferred value missing a previous value to diff",
+					"metricName", metricName, "time", currRatioReading.Time)
+				continue
+			}
+
+			value, err := diff.toValue()
+			if err != nil {
+				// The denominator did not change -- divide by zero error.
```

Comment:
[nit] Might as well report the created error here, no?

Reply:
Good idea